sklearn/feature_extraction/_hashing.pyx

```diff
@@ -103,4 +110,17 @@ def transform(raw_X, Py_ssize_t n_features, dtype, bint alternate_sign=1):
     else:
         indptr_a = indptr_a.astype(np.int32)
 
-    return (indices_a, indptr_a, values[:size])
+    return (indices_a, indptr_a, values[:size], feature_to_index_map)
+
+
+def _save_feature_mappings(into_dict, key, value):
+    """This function emulates a defaultdict with a list
+    as default value. It is used to save the mapping of
+    feature names and their indices in the value array."""
+
+    if key in into_dict:
+        into_dict[key].append(value)
+    else:
+        new_list = list()
+        new_list.append(value)
```

Comment:
this would append one value each time the word "the" is seen. That's not acceptable. It would be more appropriate to build the mapping the other way around: value => bucket, and then reverse it later.

Reply:
How about I use a set instead?

Reply:
That would avoid doing the reversing

Reply:
True, but a set is essentially a dict. That means 2 dict lookups per feature, rather than 1, and much more memory.

Reply:
The reversal has complexity as a function of the vocabulary size; this has a function of the number of words in the text. The former is asymptotically much cheaper.